### PR TITLE
TileData + Worker refactor

### PR DESCRIPTION
**Pull request by jfirebaugh**
Cleanly separate two different kinds of data:

| Map Tile Data | Worker Tile Data |
| ------ | ------ | 
| Affiliated with the Map thread | Affiliated with one of the Worker threads |
| May include GL objects | "Plain old data" only; no GL objects (#926) |
| Owned by a `Tile` object | Owned by a `TileWorker` object |
| | May be _moved_ to map thread at key moments (e.g. after a tile parse completes) |

Data will move between map thread and worker threads via the `Thread::invoke` message passing system. No tile data will be shared between the two. (For now we'll continue to share references to `Style` and atlases, but this is another step toward removing that sharing as well.)

Targeted bugs:

* https://github.com/mapbox/mapbox-gl-native/issues/926
* https://github.com/mapbox/mapbox-gl-native/issues/667 (partially)

Plan of attack:

* Collapse `TileData` inheritance hierarchy into a fully-abstract base class and `RasterTileData`, `VectorTileData`, `LiveTileData` child classes. Note that `LiveTileData` will not inherit from `VectorTileData`. (This will introduce code duplication that will later be factored away.)
* Introduce `TileWorker` class, move worker code from `TileData` to it. This may need to be a parallel inheritance hierarchy with `RasterTileWorker`, `VectorTileWorker`, and `LiveTileWorker`. At first objects of this class will be owned by `TileData` and descendants, but eventually they will live persistently on the worker thread.
* Move the making of tile requests to `TileWorker`.
* Add smarter dispatching to `Worker`. Assign a worker thread to a tile once, and then dispatch future work requests to the same worker thread, like in gl-js.
* Migrate `TileWorker` to worker threads.

There are many known unknowns here -- best way to learn them is to start working on this and see what we run into.

### CODE REVIEW

**Comment by tmpsantos** [comment]
> These are all the types of things that I expect to migrate into TileWorker, so I'm going to try introducing that now.

Hmm, so the `TileWorker` is not exactly a `Worker`. I got a bit confused at the beginning but after reading the code is quite clear to me, but maybe we should call it something like `TileWorkerJob`.

**Comment by tmpsantos** [comment]
@jfirebaugh :+1:

**Comment by jfirebaugh** [comment]
Right, TileWorker is an interim class. Once it's fully separable from the TileData hierarchy it should merge into Worker::Impl. Then we should probably rename Worker to TileWorker.

File: src/mbgl/map/live_tile_data.cpp [hunk 1]
```diff
@@ -1,58 +1,57 @@
...
+    : VectorTileData(id_, style_, source_, angle_, collisionDebug_),
       annotationManager(annotationManager_) {
-    // live features are always ready
-    setState(State::loaded);
+    // live features are always loaded
+    state = State::loaded;
 }
 
 LiveTileData::~LiveTileData() {
-    // Cancel in most derived class destructor so that worker tasks are joined before
```
Comment (kkaefer) [inline src/mbgl/map/live_tile_data.cpp]: These comments are very helpful. Any reason to remove them?

Comment (jfirebaugh) [inline src/mbgl/map/live_tile_data.cpp]: My eventual plan was to have a flat hierarchy, with `LiveTileData` _not_ inheriting from `VectorTileData`, so that (among other things) this would become moot. If we end up doing that I'll add the comments back.

File: src/mbgl/map/live_tile_data.cpp [hunk 2]
```diff
@@ -1,58 +1,57 @@
...
-    // any member data goes away.
     cancel();
 }
 
-void LiveTileData::parse() {
-    if (getState() != State::loaded) {
-        return;
+bool LiveTileData::reparse(Worker&, std::function<void()> callback) {
+    if (parsing.test_and_set(std::memory_order_acquire)) {
+        return false;
```
Comment (kkaefer) [inline src/mbgl/map/live_tile_data.cpp]: Can we store this as a `dirty` flag so we know that we weren't able to start parsing and have to try again some time later?

File: src/mbgl/map/tile_worker.cpp [hunk 3]
```diff
@@ -1,73 +1,71 @@
...
+    const auto it = buckets.find(layer.bucket->name);
+    if (it == buckets.end()) {
+        return nullptr;
+    }
+
+    assert(it->second);
+    return it->second.get();
+}
+
+size_t TileWorker::countBuckets() const {
```
Comment (kkaefer) [inline src/mbgl/map/tile_worker.cpp]: Can we get rid of the bucket count? TileData objects don't necessarily have a notion of buckets, e.g. the Raster source only ever has one "bucket".

**Comment by jfirebaugh** [comment]
I'm currently working on moving the after callback / canceling semantics offered by `WorkTask` into the core `RunLoop` implementation -- into [`Invoker`](https://github.com/mapbox/mapbox-gl-native/blob/f4707f72e0ce7f1c40337c72ae5c0c0c8ef0efed/src/mbgl/util/run_loop.hpp#L70) specifically.

This involves adding an additional template argument to `Invoker` which is instantiated with the callback type, either `std::function<void (R)>` for some type R if the invoked method returns a result, or `std::function<void ()>` if it does not. I then want to overload the `invoke` member function with implementations for these two cases:

```
        template <std::size_t... I>
        void invoke(std::index_sequence<I...>) {
            func(std::forward<Args>(std::get<I>(params))...);
            callback();
        }

        template <std::size_t... I>
        void invoke(std::index_sequence<I...>) {
            callback(func(std::forward<Args>(std::get<I>(params))...));
        }
```

I need to use some kind of `enable_if` or [SFINAE](http://en.cppreference.com/w/cpp/language/sfinae) to resolve the overloading but I'm stuck on that. Basically I want the first definition to be used for `std::function<void ()>` and the second one for `std::function<void (R)>`.

**Comment by kkaefer** [comment]
> Assign a worker thread to a tile once, and then dispatch future work requests to the same worker thread, like in gl-js.

Why is this necessary?

**Comment by jfirebaugh** [comment]
As you pointed out in https://github.com/mapbox/mapbox-gl-native/issues/1471#issuecomment-109190596, it's probably not, and I agree.

File: src/mbgl/map/live_tile_data.cpp (hunk 2 above)
Comment (jfirebaugh) [inline src/mbgl/map/live_tile_data.cpp]: I don't fully understand partial parsing yet but I assume this is handled by the return value here and the code in `Style::handlePartialTile`.

File: src/mbgl/map/tile_worker.cpp (hunk 3 above)
Comment (jfirebaugh) [inline src/mbgl/map/tile_worker.cpp]: This is a question for @tmpsantos I think. `countBuckets` use used in the logic for emitting a "tile loaded" event.

Comment (tmpsantos) [inline src/mbgl/map/tile_worker.cpp]: `countBuckets` is used to find out if, after reparsing a partial tile, anything has changed and we need to fire a "tile data changed" event. We can get rid of it after we map pending resources to tile data, because every reparse will always be caused by a resource that the tile data will be waiting for.

**Comment by jfirebaugh** [comment]
This branch is currently hanging on headless rendering tests. It has something to do with reparsing partial tiles, but the data flow there is really hard to follow. @tmpsantos do you think you could take a look at it?

**Comment by tmpsantos** [comment]
> This branch is currently hanging on headless rendering tests. It has something to do with reparsing partial tiles, but the data flow there is really hard to follow. @tmpsantos do you think you could take a look at it?

On it.

**Comment by kkaefer** [comment]
@tmpsantos I would love the partial parsing to be refactored. I'm thinking we should create a list of all layers that need parsing when we start the parse step. In the parse thread, we iterate through that list, parse, then remove from that list when we've parsed a layer. This means that at some point, this list will be empty and we know that everything's been parsed. This is likely less error prone than using the total bucket count as a measure for whether we need to parse again. This also means that we can start removing the `state` property.

**Comment by kkaefer** [comment]
The long term goal is to split up parsing per layer so that we could ideally parse one vector tile with multiple threads. This requires making the task of parsing a layer from a tile independent of order and ideally has no concurrent access. This means we'd have to give every layer their own buffers and not do label placement until we actually have the label layers parsed.
